### PR TITLE
[provider-local] Make `Ingress`es resolvable/reachable

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:

Removes the limitation of provider-local that `Ingress` objects in the seed cluster are not reachable/resolvable.
With this, the shoot node logging feature gate can also be enabled and used with provider-local.

**Which issue(s) this PR fixes**:
Fixes one of the current [limitations of provider-local](https://github.com/gardener/gardener/blob/476a39f2b59ff5b93832bdb9cd046a6c96190b2e/docs/extensions/provider-local.md#current-limitations).

**Special notes for your reviewer**:

Thanks @rfranzke for the pairing on this topic!

**Release note**:

```feature developer
provider-local now supports `Ingress` objects in the Seed cluster and now enables using the [shoot node logging feature](https://github.com/gardener/gardener/blob/master/docs/deployment/configuring_logging.md#enable-logs-from-the-shoots-node-systemd-services).
```

### CODE REVIEW

File: example/gardener-local/kind/cluster-local.yaml [hunk 1]
```diff
@@ -4,10 +4,15 @@ nodes:
 - role: control-plane
   image: kindest/node:v1.21.1
   extraPortMappings:
+  # istio-ingressgateway
```
Comment (plkokanov) [inline example/gardener-local/kind/cluster-local.yaml:7]: Nice! These comments are really useful. Previously I had to dig into the code base to see where the ports are used.

File: docs/extensions/provider-local.md [hunk 2]
```diff
@@ -115,11 +115,11 @@ In the future, this controller can be dropped when the gardenlet no longer creat
 
 #### `Service`
 
-This controller reconciles the `istio-ingress/istio-ingressgateway` `Service` in the seed cluster if the `--apiserver-sni-enabled` flag is set to `true` (default).
-Otherwise, it reconciles the `kube-apiserver` `Service` in the shoot namespaces in the seed cluster.
+This controller reconciles `Services` of type `LoadBalancer` in the local `Seed` cluster
```
Comment (plkokanov) [inline docs/extensions/provider-local.md]: nit: 
```suggestion
This controller reconciles `Services` of type `LoadBalancer` in the local `Seed` cluster.
```

File: docs/extensions/provider-local.md [hunk 3]
```diff
@@ -115,11 +115,11 @@ In the future, this controller can be dropped when the gardenlet no longer creat
...
+Since the local Kubernetes clusters used as Seed clusters typically don't support such services, this controller sets the `.status.ingress.loadBalancer.ip[0]` to the IP of the host.
+It makes important LoadBalancer Services (e.g. `istio-ingress/istio-ingressgateway` and `garden/nginx-ingress-controller`) available to the host by setting `spec.ports[].nodePort` well-known ports that are mapped to `hostPorts` in the kind cluster configuration.
```
Comment (plkokanov) [inline docs/extensions/provider-local.md]: nit:
```suggestion
It makes important LoadBalancer Services (e.g. `istio-ingress/istio-ingressgateway` and `garden/nginx-ingress-controller`) available to the host by setting `spec.ports[].nodePort` to well-known ports that are mapped to `hostPorts` in the kind cluster configuration.
```

**Review by plkokanov — CHANGES_REQUESTED** [review]
Just some minor nits that I saw, otherwise pretty easy to review :)
I'm about to do a quick test locally.

**Comment by rfranzke** [comment]
/assign

Can you check why the e2e tests are failing, please?

**Comment by rfranzke** [comment]
FYI @vlvasilev, this might be of interested for you 😉

File: pkg/provider-local/controller/ingress/add.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
```
Comment (rfranzke) [inline pkg/provider-local/controller/ingress/add.go:1]: Please add this controller with to https://github.com/gardener/gardener/blob/2a5dacb7f89414d7ed28a98fd9688283fb3977cc/charts/gardener/provider-local/extension/values.yaml#L20

**Comment by timebertt** [comment]
@plkokanov @rfranzke I addressed your review comments.

I've also found the reason for the test failures by execing into a running test job pod (couldn't reproduce locally).
Shoots got unhealthy with something like
```
$ k -n garden-local get shoot default-7nr6k -oyaml
...
  - lastTransitionTime: "2022-05-11T11:39:49Z"
    lastUpdateTime: "2022-05-11T11:38:45Z"
    message: 'Stateful set "prometheus" is unhealthy: not enough ready replicas (0/1)'
    reason: StatefulSetUnhealthy
    status: "False"
    type: ControlPlaneHealthy
...

$ k -n shoot--local--default-7nr6k describe po prometheus-0
...
Events:
  Type     Reason            Age    From               Message
  ----     ------            ----   ----               -------
  Warning  FailedScheduling  2m39s  default-scheduler  0/1 nodes are available: 1 Too many pods.
  Warning  FailedScheduling  2m37s  default-scheduler  0/1 nodes are available: 1 Too many pods.

$ k get po -A --no-headers | grep -v Pending | wc -l
110

$ k get no gardener-local-control-plane -oyaml
...
  allocatable:
    cpu: "100"
    ephemeral-storage: 102033388Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 100Gi
    pods: "110"
  capacity:
    cpu: "100"
    ephemeral-storage: 102033388Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 100Gi
    pods: "110"
```

These scheduler events would have been helpful for debugging, if they would have get dumped into the test logs.
I think, this was skipped because the context deadline was already exceeded:
```
    {"level":"error","msg":"Cannot dump shoot state default-6vm7d: retry failed with context deadline exceeded, last error: shoot \"default-6vm7d\" was not successfully reconciled","ts":"2022-05-04T17:20:54.149Z"}
```

This is addressed as well in a new commit.
I will check again later, what's our best way to deal with the max pods limitation.

**Comment by rfranzke** [comment]
Nice finding @timebertt. We can probably use https://kind.sigs.k8s.io/docs/user/configuration/#kubeadm-config-patches to set the kubelet config in the kind cluster accordingly, right?

**Comment by timebertt** [comment]
> These scheduler events would have been helpful for debugging, if they would have get dumped into the test logs. I think, this was skipped because the context deadline was already exceeded: [...]
> This is addressed as well in a new commit. I will check again later, what's our best way to deal with the max pods limitation.

Ok nice, this worked at least. The build logs now include the full event dump for better debugging including:

```
{"level":"info","msg":"At 0001-01-01 00:00:00 +0000 UTC - event for gardener-resource-manager-79ffd6d47d-7rdzj: { } FailedScheduling: 0/1 nodes are available: 1 Too many pods.","ts":"2022-05-11T12:22:22.626Z"}
```

**Comment by timebertt** [comment]
Ok, I pushed another commit increasing `maxPods` to 500 in our kind config.

I tested this out locally and increasing `maxPods` seems to be the only thing we need to do.
We are using the default `calico-ipam` in kind, which ignores `node.spec.podCIDR` and instead manages IP blocks on its own via the `ipamblocks.crd.projectcalico.org` CR. It uses `/26` blocks by default, but creates new pools as needed.

```
$ k get no gardener-local-control-plane -oyaml | grep -e podCIDR: -e pods:
  podCIDR: 10.1.0.0/24
    pods: "500"
    pods: "500"

$ k get deploy nginx # we have 10 system pods
NAME    READY     UP-TO-DATE   AVAILABLE   AGE
nginx   490/500   500          490         16m

$ k get ipamblocks.crd.projectcalico.org
NAME              AGE
10-1-130-192-26   16m
10-1-131-0-26     15m
10-1-131-128-26   14m
10-1-131-192-26   14m
10-1-131-64-26    14m
10-1-132-0-26     14m
10-1-132-128-26   13m
10-1-132-64-26    13m
```

Let's see how this works out in CI 🚀

**Comment by rfranzke** [comment]
After a joint review with @oliver-goetz:

/lgtm
/approve

**Pull request merged**